### PR TITLE
chore(gatsby): Migrate type of sort to TypeScript

### PR DESCRIPTION
**Pull request by purelledhand**
## Description

Migrate `package/src/type/sort.js` to TypeScript

## Related Issues

Related to #21995

### CODE REVIEW

File: packages/gatsby/src/schema/types/sort.ts [hunk 1]
```diff
@@ -1,45 +1,79 @@
 // @flow
```
Comment (pieh) [inline packages/gatsby/src/schema/types/sort.ts:1]: Let's remove this line - it's no longer flow typed

File: packages/gatsby/src/schema/types/sort.ts [hunk 2]
```diff
@@ -100,10 +52,10 @@ const convert = (
       }
     }
 
-    const type = getNamedType(fieldConfig.type)
+    const type = fieldConfig.type
     if (type instanceof GraphQLInputObjectType) {
       if (depth < MAX_SORT_DEPTH) {
-        const typeComposer = schemaComposer.getAnyTC(
+        const typeComposer = schemaComposer.getIFTC(
```
Comment (pieh) [inline packages/gatsby/src/schema/types/sort.ts:58]: should this be getITC and not getITFC?

getITC is for inputs and getITFC is for interfaces

I didn't look up code around it much yet, just the type instanceof GraphQLInputObjectType part

File: packages/gatsby/src/schema/types/sort.ts [hunk 3]
```diff
@@ -1,84 +1,36 @@
...
+} from "graphql"
+import { addDerivedType } from "./derived-types"
+import {
+  InputTypeComposer,
+  SchemaComposer,
+  EnumTypeComposer,
+  InterfaceTypeComposer,
+} from "graphql-compose"
+import { ComposeEnumValueConfig } from "graphql-compose/lib/EnumTypeComposer"
+import { ObjMap } from "graphql-compose/lib/utils/definitions"
```
Comment (pieh) [inline packages/gatsby/src/schema/types/sort.ts:15]: Shouldn't this be available from `graphql-compose` main endpoint - given that `utils/definitions` are re-exported in https://github.com/graphql-compose/graphql-compose/blob/29b237c3f556891fa7cd025500dddeb1cdc12ffd/src/index.d.ts#L37

File: packages/gatsby/src/schema/types/sort.ts (hunk 2 above)
Comment (purelledhand) [reply packages/gatsby/src/schema/types/sort.ts:58]: `typeComposer` in this function is invoke `getFieldExtension` getter method like below line.
`typeComposer.getFieldExtension(fieldName, "sortable")`
and this customized extension field is defined in `nodeInterfaceTC: IFTC` at `./node-interface.js` like below code.

```
extensions: {
          searchable: SEARCHABLE_ENUM.SEARCHABLE,
          sortable: SORTABLE_ENUM.SORTABLE,
          needsResolve: true,
        }
```

I would appreciate any better way🙏🏻

File: packages/gatsby/src/schema/types/sort.ts (hunk 3 above)
Comment (pieh) [reply packages/gatsby/src/schema/types/sort.ts:15]: ah, we are on older major :S ok, makes sense, tho I don't like importing internals like that :(

File: packages/gatsby/src/schema/types/sort.ts (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/schema/types/sort.ts:58]: I am out of my depth here. I will need to leave proper review for our GraphQL folks

File: packages/gatsby/src/schema/types/sort.ts (hunk 3 above)
Comment (purelledhand) [reply packages/gatsby/src/schema/types/sort.ts:15]: Totally agree with you. I'll edit it!

**Comment by purelledhand** [comment]
> Thanks for the PR! But unit tests are failing on it.
> 
> As a general suggestion, I recommend not to change the logic because most of those changes seem wrong (and thus fail the tests).
> 
> Maybe as a first step, we should just add types where possible and use `any` `@ts-ignore` for things that do not play well with the existing code.

Thank you for review! I recovered logic and passed all unit test, but update_i18n_source test is failed. because CLI required password of 'undefined@github.com' while test like below code.
Can I get some advice on this issue?
```bash
[2020-05-15T04:49:25.676] [INFO] update-source - Pushing changes to gatsby-i18n-source
Password for 'https://undefined@github.com':  
 
Too long with no output (exceeded 10m0s): context deadline exceeded 
```

> Also, is there a reason to move a bunch of functions to the bottom of the file? It makes reviewing very inconvenient.

And the reason to move `convert` function to the top of the file is that otherwise, the lint test fail due to the lint warning that it was used before declaring the function.

**Comment by vladar** [comment]
@purelledhand somehow I missed your last message, sorry for that. In the meantime, this was ported to TS via #26369

**Pull request closed without merging**